automation/sheet_checker.py

```diff
@@ -0,0 +1,165 @@
+        self.logger = logging.getLogger(__name__)
+        timestamp = datetime.now().strftime('%Y-%m-%d')
+        logging.basicConfig(filename='%s+%s.log' % (self.dataset, timestamp), level=logging.INFO)
+
+    def run(self, force_full_update=False):
+        if self.is_sheet_updated() or force_full_update:
+            self.logger.info("Sheet has been updated since last run. Cleaning and compressing...")
+            self.set_up_environment()
+            self.clean()
+            self.compress()
+            self.update_last_ran_ts()
+            self.clean_up_environment()
+        else:
+            self.logger.info("Sheet has not been updated since last run. Exiting.")
+            self.update_last_ran_ts()
+
+    def clean(self):
+        try:
+            for cleaning_nb in self.cleaning_nbs:
+                self.run_cleaning_notebook(cleaning_nb)
+            emailer.send_success_email("Cleaning", self.dataset)
+        except Exception as e:
+            self.logger.exception(e)
+            emailer.send_fail_email("Cleaning", self.dataset)
+            self.logger.warning('Cleaning failed.')
+            sys.exit('Exiting: encountered an issue while cleaning.')
+
+    def compress(self):
+        try:
+            self.run_compression_notebook()
+            emailer.send_success_email("Compressing", self.dataset)
+        except Exception as e:
+            self.logger.exception(e)
+            self.logger.warning('Compressing failed.')
+            emailer.send_fail_email("Compressing", self.dataset)
+            sys.exit('Exiting: encountered an issue while compressing.')
+        self.logger.info("Successfully cleaned and compressed data.")
+
+    def run_cleaning_notebook(self, cleaning_nb_name):
+        out_notebook_name = 'cleaning_%s_result_nb.ipynb' % self.dataset
+        nb = nbformat.read('../data_cleaning/%s' % cleaning_nb_name, as_version=4)
+        ep = ExecutePreprocessor(timeout=600, kernel_name='python3')
+        self.run_notebook(ep, nb, out_notebook_name)
+        self.logger.info("Successfully cleaned.")
+
+    def run_compression_notebook(self):
+        out_notebook_name = 'compression_%s_result_nb.ipynb' % self.dataset
+        nb = nbformat.read('../data_cleaning/create_datasets_for_website.ipynb', as_version=4)
+        ep = ExecutePreprocessor(timeout=600, kernel_name='python3')
+        self.run_notebook(ep, nb, out_notebook_name)
+        self.logger.info("Successfully compressed.")
+        return True
+
+    def run_notebook(self, ep, nb, out_notebook_name):
+        try:
+            ep.preprocess(nb, {'metadata': {'path': '../data_cleaning/'}})
+        except CellExecutionError:
+            msg = 'Error executing the notebook'
+            msg += 'See notebook "%s" for the traceback.' % out_notebook_name
+            self.logger.info(msg)
+            raise
+        finally:
+            with open(out_notebook_name, mode='wt') as f:
+                nbformat.write(nb, f)
```

Comment:
To check my understanding -- the `ExecutionPreprocessor` runs the notebook in memory, but doesn't change the original file? Hence you need to run `nbformat.write` here?

Reply:
Right, it runs the nb and saves the result in a different location (out_notebook_name)